controllers/kvmcluster_controller.go

```diff
@@ -0,0 +1,159 @@
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the KVMCluster object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *KVMClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reconcileError error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	defer metrics.CaptureMetrics()
+
+	// Fetch the KVMCluster instance
+	kvmCluster := &v1alpha4.KVMCluster{}
+	err := r.Get(ctx, req.NamespacedName, kvmCluster)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	defer func() {
+		if reconcileError != nil {
+			r.Recorder.Event(kvmCluster, "Warning", "failed-to-reconcile", reconcileError.Error())
+		}
+	}()
+
+	// Fetch the Cluster.
+	cluster, err := util.GetOwnerCluster(ctx, r.Client, kvmCluster.ObjectMeta)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if cluster == nil {
+		log.Info("Cluster Controller has not yet set OwnerRef")
+		return reconcile.Result{}, nil
+	}
+
+	if annotations.IsPaused(cluster, kvmCluster) {
+		log.Info("KVMCluster or linked Cluster is marked as paused. Won't reconcile")
+		return reconcile.Result{}, nil
+	}
+
+	log = log.WithValues("cluster", cluster.Name)
+
+	// Create the scope.
+	clusterScope, err := scope.NewClusterScope(scope.ClusterScopeParams{
+		Cluster:        cluster,
+		KVMCluster:     kvmCluster,
+		ControllerName: KVMClusterController,
+	})
+	if err != nil {
+		return reconcile.Result{}, errors.Errorf("failed to create scope: %+v", err)
+	}
+
+	// Always close the scope when exiting this function so we can persist any KVMCluster changes.
+	defer func() {
+		if err := clusterScope.Close(); err != nil && reconcileError == nil {
+			reconcileError = err
+		}
+	}()
+
+	// Handle deleted clusters
+	if !kvmCluster.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, clusterScope)
+	}
+
+	// Handle non-deleted clusters
+	return r.reconcileNormal(ctx, clusterScope)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KVMClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha4.KVMCluster{}).
+		Complete(r)
+}
+
+func (r *KVMClusterReconciler) reconcileNormal(ctx context.Context, scope *scope.ClusterScope) (_ reconcile.Result, reconcileError error) {
+
+	// If the KVMCluster doesn't have our finalizer, add it.
+	controllerutil.AddFinalizer(scope.KVMCluster, v1alpha4.ClusterFinalizer)
```

Comment:
Full control over the finalizers and nice utils to do that (5. [here](https://github.com/giantswarm/giantswarm/issues/17771#issuecomment-867471192)).

Reply:
We might only want to add the finalizer and patch if the finalizer doesn't exist. Is that already handled by `AddFinalizer` and `scope.PatchObject()`?

Reply:
`AddFinalizer` only adds the finalizer if it's not already there: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/controllerutil.go#L351.

As of `scope.PatchObject()` - it falls back to a `patchHelper` ([upstream utils](https://github.com/kubernetes-sigs/cluster-api/blob/master/util/patch/patch.go#L83)) that [checks](https://github.com/kubernetes-sigs/cluster-api/blob/master/util/patch/patch.go#L273-L275) if the objects changed before actually patching it.